src/mbgl/tile/geometry_tile_worker.hpp

```diff
@@ -85,11 +94,21 @@ class GeometryTileWorker {
     optional<std::unique_ptr<const GeometryTileData>> data;
 
     bool symbolLayoutsNeedPreparation = false;
+    bool patternNeedsLayout = false;
+
     std::vector<std::unique_ptr<SymbolLayout>> symbolLayouts;
+
+    using LinePatternLayout = PatternLayout<LineBucket>;
+    using FillPatternLayout = PatternLayout<FillBucket>;
+    using FillExtrusionPatternLayout = PatternLayout<FillExtrusionBucket>;
+
+    std::vector<variant<std::unique_ptr<LinePatternLayout>, std::unique_ptr<FillPatternLayout>, std::unique_ptr<FillExtrusionPatternLayout>>> patternLayouts;
```

Comment:
This is an instance where we'd be better off with an abstract `Layout` base class rather than using `variant`. Both `SymbolLayout` and `PatternLayout` should inherit from it. The goal should be that `GeometryTileWorker` can treat layouts uniformly:

* Replace `symbolLayouts` and `patternLayouts` with a single `std::vector<std::unique_ptr<Layout>> layouts` member, and a single loop over all `layouts` in `performSymbolLayout`.
* `GeometryTileWorker` has two steps, currently named `parse` and `performSymbolLayout`. Let's come up with better names. Perhaps `prepareLayout` and `finalizeLayout`? While we're at it, we should also take the `Symbol` out of `hasPendingSymbolDependencies`. Symbols aren't so special anymore.
* Make `createLayout` a virtual method in `RenderLayer`. It can return a null `unique_ptr<Bucket>`, indicating that `createBucket` should be used instead. (Or should we make things completely uniform, and have a simple `Layout` subclass for every layer type, even those that don't have dependencies? Then we could call the two worker steps `createLayouts` and `createBuckets`, and have only a single `RenderLayer::createLayout` method. This approach would be more work, but sounds cleaner.)

You'll need to do some design work to figure out what the `Layout` interface should look like. It looks like it might be a single method `createBucket` (name open for discussion):

* For the concrete subclass `SymbolLayout`, this would be a combination of the methods `prepare` and `place`.
* For pattern layouts, there's already a `createBucket` method. Is it possible to remove the conditional in `checkPatternLayout`, always create the layout, and have it do the work that `checkPatternLayout` is doing now in the case that a pattern isn't being used?
* To avoid `GeometryTileWorker` needing to iterate over concrete `layerPaintProperties` types, it's fine if `createBucket` takes as a mutable input parameter the `buckets` map and emplaces into it in a loop itself.
* `createBucket` will need a lot of input parameters, not all of which are used by every subclass. That's okay. One pattern that's been effective in situations like this is to package all the parameters into a struct (`BucketParameters`, `PropertyEvaluationParameters`, etc.). This allows us to add or remove arguments without needing to modify the parameter list in every subclass.

I think these changes will make `GeometryTileWorker` easier to understand and modify, and hopefully it will help reduce the binary size impact as well.